blueprints/ort-rewrite-unix-style.md

```diff
@@ -0,0 +1,210 @@
+    - INPUT: configuration and specification to fetch and emplace config files.
+    - BEHAVIOR: fetches and places config files
+    - OUTPUT: success or failure message
+
+- **Traffic Ops Requestor**. This will fetch data needed from Traffic Ops, such as the Update Pending flag, packages, etc. This should never modify TO data, and should be guaranteed read-only. Any status modifications should go in the Traffic Ops Updater.
+    - INPUT: Traffic Ops URL and credentials, and data to fetch
+    - BEHAVIOR: Requests data from Traffic Ops
+    - OUTPUT: Traffic Ops data requested
+        - Format is probably multipart/mixed, but format may be different if a better format is determined. Ideal "UNIX Philosophy" format is line-delimited text, but the complexity may preclude that. The more complex and difficult to parse, the further from the "UNIX Philosophy." E.g. multipart/mixed is preferable to JSON.
+- **Config File Generator**. This will take TO data and produce config files.
+    - INPUT: Traffic Ops data, and config file(s) to generate or exclude (typically all, possibly “reval only” or other behaviors of ORT)
+    - BEHAVIOR: No side effects. Computationally: builds requested files.
+    - OUTPUT: Config files
+        - Format is probably multipart/mixed, but as above, may be different, multipart is preferable to JSON, etc.
+- **Config File Preprocessor**. Preprocesses generated config files, making post-generation modifications such as underscore directive replacements, and remap OVERRIDE replacements. TODO: determine if this should be rolled into the Config File Generator.
+    - INPUT: Config files, TO data/metadata.
+    - BEHAVIOR: No side effects; computationally: modifies files per rules and data.
+    - OUTPUT: modified config files.
+- **Server Config Readiness Verifier**. Verifies the operating system is ready and safe to apply the given config. Currently, this is just checking udev rules and verifying ATS block devices don’t have filessytems. But it may be more in the future. TODO: determine if necessary; should this even be ORT’s job? Will we ever need anything besides udev/50-ats.rules? 
+    - INPUT: config files
+    - BEHAVIOR: No side effects; reads configs and inspects server state.
+    - OUTPUT: whether server is safe and ready to apply config files.
+- **ATS Plugin Readiness Verifier**. Verifies ATS has the necessary plugins for the config files.
+    - INPUT: config files
+    - BEHAVIOR: No side effects; reads configs and inspects server state.
+    - OUTPUT: whether ATS has all necessary plugins, or which files require which missing plugins.
+- **Diff Tool**. This will take two config files (presumably an existing and new file) and return their differences. TODO: determine if this is necessary; if configs are deterministic, can POSIX diff be used?
+    - INPUT: Two config files
+    - BEHAVIOR: No side effects. Computationally: diffs given files.
+    - OUTPUT: file diff. Ideally in a standard format.
```

Comment:
Even if configs are deterministic, a diff output for humans will be very valuable. And POSIX diff on things like YAML is basically useless.

Reply:
Not opposed to a smarter diff. But: 

>POSIX diff on things like YAML is basically useless.

Why? YAML is line-based. And even if a config weren't, the operator can and should format their code on lines, so the diff does make sense. I can't imagine a config needing to be minified

Reply:
YAML diffs being useless feels like it's based on the assumption that config outputs are _not_ deterministic. Which I hope won't be the case.

Currently the only reason POSIX diff doesn't work is because (besides lingering non-determinism) the files all come with that header that says "generated by x on y date". I hope we can reduce that to just `generated by x` now because the file modification time should tell you the "when", and that'll mean the output is truly only dependent on TO state (which is arguably an input so "stateless" may apply). So all non-zero POSIX diffs would be real.

I guess what I'm saying is: whether or not POSIX diff can be used instead comes down to the implementation of the configuration file generator. If we can declare in this blueprint that it be deterministic and change the header format a bit, then we can say that POSIX diff can be used.